docs/build-speed-android.md

```diff
@@ -0,0 +1,132 @@
+---
+id: build-speed-android
+title: Speeding up your Android Build
+---
+
+Building your React Native app on Android could be **expensive** and take several minutes of developers time.
+This can be problematic as your project grows and generally in bigger organizations with multiple React Native developers.
+
+With [the New React Native Architecture](/docs/next/new-architecture-app-modules-android), this problem is becoming more critical
+as you might have to compile some native code (C++) in your project with the Android NDK.
+
+To mitigate this performance hit, this page shares some suggestions on how to **improve your Android build time**.
+
+## Build only one ABI
+
+When building your app locally, you build all the 4 ABIs by default: `armeabi-v7a`, `arm64-v8a`, `x86` & `x86_64`.
+
+However, you probably don't need to build all of them if you're building locally and testing your emulator or on a phisical device.
+
+This should reduce your build time by a **~75% factor**.
+
+If you're using the React Native CLI, you can use the `--active-arch-only` flag together with the `run-android` command.
+This flag will make sure the correct ABI is picked up from either the running emulator or the plugged in phone.
+To confirm that this approach is working fine, you'll see a message like `info Detected architectures arm64-v8a` on console.
+
+```
+$ yarn react-native run-android --active-arch-only
+
+[ ... ]
+info Running jetifier to migrate libraries to AndroidX. You can disable it using "--no-jetifier" flag.
+Jetifier found 1037 file(s) to forward-jetify. Using 32 workers...
+info JS server already running.
+info Detected architectures arm64-v8a
+info Installing the app...
+```
+
+This mechanism relies on the `reactNativeArchitectures` Gradle property.
+
+Therefore, if you're building directly with Gradle from the command line and without the CLI, you can specify the ABI you want to build as follows:
+
+```
+$ ./gradlew :app:assembleDebug -PreactNativeArchitectures=x86,x86_64
+```
+
+This can be useful if you wish to build your Android App on a CI and use a matrix to parallelize the build of the different architectures.
+
+If you wish, you can also override this value locally, using the `gradle.properties` file you have in the [top level folder](https://github.com/facebook/react-native/blob/19cf70266eb8ca151aa0cc46ac4c09cb987b2ceb/template/android/gradle.properties#L30-L33) of your project:
+
+```
+# Use this property to specify which architecture you want to build.
+# You can also override it from the CLI using
+# ./gradlew <task> -PreactNativeArchitectures=x86_64
+reactNativeArchitectures=armeabi-v7a,arm64-v8a,x86,x86_64
```

Comment:
i was giving this a shot for speediness

```diff
reactNativeArchitectures=arm64-v8a
```

but keep getting

```
[CXX1405] error when building with ndkBuild using /Users/leo/Documents/GitHub/react-native-template-typescript/android/app/src/main/jni/Android.mk: Build command failed.
Error while executing process /Users/leo/Library/Android/sdk/ndk/24.0.8079956/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/src/main/jni/Android.mk APP_ABI=armeabi-v7a NDK_ALL_ABIS=armeabi-v7a NDK_DEBUG=1 NDK_OUT=/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/build/intermediates/cxx/Debug/6x5h664i/obj NDK_LIBS_OUT=/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/build/intermediates/cxx/Debug/6x5h664i/lib APP_CFLAGS+=-Wall APP_CFLAGS+=-Werror APP_CFLAGS+=-fexceptions APP_CFLAGS+=-frtti APP_CFLAGS+=-DWITH_INSPECTOR=1 APP_CPPFLAGS+=-std=c++17 APP_PLATFORM=android-21 APP_STL=c++_shared NDK_TOOLCHAIN_VERSION=clang GENERATED_SRC_DIR=/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/build/generated/source PROJECT_BUILD_DIR=/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/build REACT_ANDROID_DIR=/Users/leo/Documents/GitHub/react-native-template-typescript/android/../node_modules/react-native/ReactAndroid REACT_ANDROID_BUILD_DIR=/Users/leo/Documents/GitHub/react-native-template-typescript/android/../node_modules/react-native/ReactAndroid/build APP_SHORT_COMMANDS=false LOCAL_SHORT_COMMANDS=false -B -n}
Android NDK: ERROR:/Users/leo/Documents/GitHub/react-native-template-typescript/android/app/src/main/jni/Android.mk:fb: LOCAL_SRC_FILES points to a missing file    
Android NDK: Check that /Users/leo/Documents/GitHub/react-native-template-typescript/android/app/build/react-ndk/exported/armeabi-v7a/libfb.so exists  or that its path is correct   

fcntl(): Bad file descriptor
/Users/leo/Library/Android/sdk/ndk/24.0.8079956/build/core/prebuilt-library.mk:51: *** Android NDK: Aborting    .  Stop.
```

which gets resolved and builds successfully when i include all 4 ABI's

noting `armeabi-v7a` missing from the error msg

i can get a repro branch/draftPR up in my template if helps

Reply:
> i can get a repro branch/draftPR up in my template if helps

That can help. It seems like this configuration is not picked up correctly.

Reply:
tl;dr lgtm, it's jus me, got this weird issue where the first build fails with same msg, then subsequent succeed

[minimal repro](https://github.com/leotm/react-native-template-typescript)
_removed: react-native-screens, react-native-safe-area-context_
- `git clean -fdx`
- `yarn`
- [Create `local.properties` with NDK `24.0.8079956`](https://github.com/leotm/react-native-template-typescript#android)
- _Forgot I left Gradle JDK default on 11.0.12_
- Android Studio > Make Project > Failed (1min)
- Android Studio > Make Project > Finished (4m)
- Android Studio > Make Project > Finished (24-25s since cached)

```diff
- reactNativeArchitectures=armeabi-v7a,arm64-v8a,x86,x86_64
+ reactNativeArchitectures=armeabi-v7a,arm64-v8a,x86_64
# working (sync, make, 20-34s, cached)
```

```diff
- reactNativeArchitectures=armeabi-v7a,arm64-v8a,x86_64
+ reactNativeArchitectures=arm64-v8a,x86_64
# working (sync, 18-21s, cached)
```

```diff
- reactNativeArchitectures=arm64-v8a,x86_64
+ reactNativeArchitectures=arm64-v8a
# working (sync, 18-20s, cached)
```

```
reactNativeArchitectures=arm64-v8a
# temurin-11
# rebuild, make project fails, then working 2m30)
```

```
reactNativeArchitectures=arm64-v8a
# azul-11 aarch64
# rebuild, make project fails, then working 1m34s)
```

keep up the good work

Reply:
Thanks for the extensive testing 👍

Reply:
I can confirm that this is a "bug". I'll push a fix for this soon-ish. Practically, `--active-arch-only` will not work on a fully-clean environment. You need to do a full build before. There is a simple workaround which I'll add to this doc in a follow-up PR